walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,118 @@
+        if (checkMandatoryAttendee && MandatoryAttendants.contains(attendee)) {
+            handleEventConflict(mandatoryTimeRanges, e, request);
+            checkMandatoryAttendee = false;
+        }
+        if (optionalAttendants.contains(attendee)) {
+            numOptionalUnavailable++;
+        }
+      }
+      //if all optional attendees are attending this event, then mark this time slot as unavailable for any optional attendees
+      if (optionalAttendants.size() != 0 && numOptionalUnavailable == optionalAttendants.size()) {
+          handleEventConflict(optionalTimeRanges, e, request);
+      }
+    }
+
+    // See if there are any valid overlaps between the mandatory time ranges and optional time ranges
+    ArrayList<TimeRange> overlapTimeRanges = new ArrayList<TimeRange>();
+    int mandatoryIndex = 0;
+    int optionalIndex = 0;
+    while (mandatoryIndex < mandatoryTimeRanges.size() && optionalIndex < optionalTimeRanges.size()) {
+        TimeRange curMandatoryRange = mandatoryTimeRanges.get(mandatoryIndex);
+        TimeRange curOptionalRange = optionalTimeRanges.get(optionalIndex);
+        // if they overlap, need to see if any common timerange can be used
+        TimeRange overlap = getOverlap(curMandatoryRange, curOptionalRange);
+        if (overlap != null && overlap.duration() >= request.getDuration()) {
+            overlapTimeRanges.add(overlap);
+        } 
+        if (curOptionalRange.end() <= curMandatoryRange.end()) {
+            optionalIndex++;
+        }
+        if (curMandatoryRange.end() <= curOptionalRange.end()) {
+            mandatoryIndex++;
+        }
+    }
+    if (MandatoryAttendants.size() != 0 && overlapTimeRanges.size() == 0) {
+        return mandatoryTimeRanges;
+    }    
+    return overlapTimeRanges;
   }
+
+    private TimeRange getOverlap(TimeRange mandatoryRange, TimeRange optionalRange) {
+        if (!mandatoryRange.overlaps(optionalRange)) { 
+            return null;
+        }
+        // if the optional range contains the mandatory range, then the entirety of the mandatory range is shared by all
+        if (optionalRange.contains(mandatoryRange)) {
+            return mandatoryRange;
+        }
+        // if the mandatory range contains the optional range, the entirety of optional range is shared by all
+        if (mandatoryRange.contains(optionalRange)) {
+            return optionalRange;
+        }
+        if (mandatoryRange.start() < optionalRange.start()) {
+            return TimeRange.fromStartEnd(optionalRange.start(), mandatoryRange.end(), true);
+        } else {
+            return TimeRange.fromStartEnd(mandatoryRange.start(), optionalRange.end(), true);
+        }
+    }
+
+    private void handleEventConflict(ArrayList<TimeRange> timeranges, Event e, MeetingRequest request) {
+        for (int i = 0; i < timeranges.size();) {
```

Comment:
So what you're doing here is iterating over indices and deleting items. I believe this is going to be problematic because as you remove items, the list gets shorter and the indices mean different things since ArrayLists are dynamic.

As an example, I think this snippet:
```
ArrayList<String> x = new ArrayList<String>(Array.asList("hi", "bye"));
for (int i = 0; i < x.size();) {
  System.out.println(i);
  System.out.println(x[i]);
  x.remove(i);
  System.out.println(x[i]);
}
```

Would output:
```
0
hi
bye
```

Definitely double check this belief since Java is not my best language. Here's a discussion I used as a source that might be helpful if you need to do something else: https://stackoverflow.com/questions/10431981/remove-elements-from-collection-while-iterating

Reply:
Yes you're correct! I believe I handled this situation however..which is why I don't automatically indent i (i++ in the for loop header) and instead manually increment i as necessary. Perhaps this isn't best code practice?

Reply:
Yeah it might not be the best practice. Admittedly I was pretty tired when I read this CL, but I found it a bit confusing. That kind of deletion sends up a red flag to a reader because it's a common mistake. Even if you've mitigated it and everything works fine, you don't want to use patterns that read as a bug to a casual observer. Think about using a different deletion pattern.

Reply:
Sounds good, just implemented an alternative solution.